crates/astria-composer/src/composer.rs

```diff
@@ -232,26 +312,166 @@ impl Composer {
+    api_server_task_handle: Option<JoinHandle<eyre::Result<()>>>,
+    executor_task_handle: Option<JoinHandle<eyre::Result<()>>>,
+    grpc_server_task_handle: Option<JoinHandle<eyre::Result<()>>>,
+    geth_collector_tasks: JoinMap<String, eyre::Result<()>>,
+}
+
+impl ShutdownInfo {
+    async fn run(self) -> eyre::Result<()> {
+        let Self {
+            composer_shutdown_token,
+            api_server_shutdown_token,
+            api_server_task_handle,
+            executor_task_handle,
+            grpc_server_task_handle,
+            mut geth_collector_tasks,
+        } = self;
+
+        // if the composer is shutting down because of an unexpected shutdown from any one of the
+        // components(and not because of a SIGTERM), we need to send the cancel signal to all
+        // the other components.
+        composer_shutdown_token.cancel();
+        // k8s issues SIGKILL in 30s, so we need to make sure that the shutdown happens before 30s.
+
+        // We give executor 17 seconds to shut down. The logic to timeout is in the
+        // executor itself. We wait 17s for all the bundles to be drained.
+        if let Some(executor_task_handle) = executor_task_handle {
+            match executor_task_handle.await {
+                Ok(res) => match res {
+                    Ok(()) => {
+                        info!("executor shut down");
+                    }
+                    Err(err) => {
+                        error!(%err, "executor shutdown with error");
+                    }
+                },
+                Err(err) => {
+                    error!(%err, "executor task didn't complete successfully");
+                }
+            }
+        } else {
+            info!("executor task was already dead");
+        };
+
+        // We give the grpc server 5 seconds to shutdown.
+        if let Some(grpc_server_task_handle) = grpc_server_task_handle {
+            match tokio::time::timeout(
+                std::time::Duration::from_secs(GRPC_SERVER_SHUTDOWN_TIME),
+                grpc_server_task_handle,
+            )
+            .await
+            .map(flatten)
+            {
+                Ok(Ok(())) => info!("grpc server shut down"),
+                Ok(Err(error)) => error!(%error, "grpc server shutdown with error"),
+                Err(error) => error!(%error, "grpc server failed to shut down in time"),
+            }
+        } else {
+            info!("grpc server task was already dead");
+        };
+
+        let geth_collectors_tasks_no = geth_collector_tasks.len();
+        let shutdown_loop = async {
+            while let Some((name, res)) = geth_collector_tasks.join_next().await {
+                let message = "task shut down";
+                match flatten(res) {
+                    Ok(()) => info!(name, message),
+                    Err(error) => error!(name, %error, message),
+                }
+            }
+        };
+
+        if geth_collectors_tasks_no > 0 {
```

Comment:
IMO this is unnecessary information. Just timeout the shutdown loop without checking the number of collector tasks.